### ISSUE TITLE
Cannot use both TypeGuard and NoReturn in return type for same `@overload`

### ISSUE THREAD

**Issue by davidfstr** [microsoft/pyright#3216]
**Describe the bug**
It appears that you cannot have a pair of `@overload` cases where one returns a `TypeGuard[T]` and the other returns a `NoReturn`.

**To Reproduce**

* With the following two files, with Python 3.8 installed:

```
# pyproject.toml
[tool.pyright]
include = ["."]
pythonPlatform = "All"

# trycast_mini.py
from typing import get_args, get_origin, Literal, NoReturn, overload, Type, TypeVar, Union
from typing_extensions import TypeGuard

_T = TypeVar("_T")

@overload
def isassignable(value: object, tp: str, *, eval: Literal[False]) -> NoReturn: ...
@overload
def isassignable(value: object, tp: Type[_T], *, eval: bool = True) -> TypeGuard[_T]: ...

def isassignable(value, tp, *, eval=True):
    if isinstance(tp, str):
        if not eval:
            raise ValueError("Cannot use a string-based type when eval=False")
        else:
            raise NotImplementedError()
    
    if get_origin(tp) is Union:
        if get_args(tp) == (int, str):
            return isinstance(value, int) or isinstance(value, str)
        raise NotImplementedError()
    return isinstance(value, tp)
```

* ...run the following commands:
    * `pip3 install pyright==1.1.230`
    * `pyright`

**Expected behavior**
There should be no typechecking errors.

**Actual behavior**

Pyright gives the following error:
```
/Users/me/tmp/trycast_mini.py
  /Users/me/tmp/trycast_mini.py:11:5 - error: Overloaded function implementation is not consistent with signature of overload 2
    Function return type "TypeGuard[_T@isassignable]" is incompatible with type "bool"
      "TypeGuard[_T@isassignable]" is incompatible with "bool" (reportGeneralTypeIssues)
1 errors, 0 warnings, 0 informations
```

Mypy gives no error on the same code.

**Additional context**
Similar to the problem where [I cannot use `TypeGuard[T]` and `bool` together](https://github.com/microsoft/pyright/issues/3215), it appears I cannot use `TypeGuard[T]` and `NoReturn` together.

**Comment by erictraut** [comment]
This works fine if you provide the appropriate type annotations for the implementation:

```python
def isassignable(value, tp: str | Type[_T], *, eval=True) -> TypeGuard[_T]:
```

**Comment by davidfstr** [comment]
For future readers, I fixed this problem while preserving the `@overload` by using a `cast(TypeGuard[_T], ...)` on every appropriate return statement:

```python
# trycast_mini.py
from typing import cast, get_args, get_origin, Literal, NoReturn, overload, Type, TypeVar, Union
from typing_extensions import TypeGuard

_T = TypeVar("_T")

@overload
def isassignable(value: object, tp: str, *, eval: Literal[False]) -> NoReturn: ...
@overload
def isassignable(value: object, tp: Type[_T], *, eval: bool = True) -> TypeGuard[_T]: ...

def isassignable(value, tp, *, eval=True):
    if isinstance(tp, str):
        if not eval:
            raise ValueError("Cannot use a string-based type when eval=False")
        else:
            raise NotImplementedError()
    
    if get_origin(tp) is Union:
        if get_args(tp) == (int, str):
            return cast(
                TypeGuard[_T],
                isinstance(value, int) or isinstance(value, str)
            )
        raise NotImplementedError()
    return cast(
        TypeGuard[_T],
        isinstance(value, tp)
    )

reveal_type(isassignable("hello", "str", eval=False))  # does reveal `NoReturn` 👍 
print("This print should be unreachable")

```
